src/dataflow/src/render/reduce.rs

```diff
@@ -523,6 +540,50 @@ where
         (key, values)
     });
 
+    if monotonic {
+        use differential_dataflow::operators::consolidate::ConsolidateStream;
+        use timely::dataflow::operators::Map;
+
+        let collection = collection
+            .consolidate()
+            .inner
+            .map(move |((key, values), time, diff)| {
+                assert!(diff > 0);
+                let aggr_funcs = aggr_funcs.clone();
+                let mut output = Vec::new();
+                for (row, func) in values.into_iter().zip(aggr_funcs.iter()) {
+                    output.push(monoids::get_monoid(row, func).expect(
+                        "hierarchical aggregations are expected to have monoid implementations",
+                    ));
+                }
+
+                ((key, ()), time, DiffVector::new(output))
+            })
+            .as_collection();
+        let arrangement = collection
+            .consolidate_stream()
+            .reduce_abelian::<_, OrdValSpine<_, _, _, _>>("ReduceMonotonicHierarchical", {
```

Comment:
I haven't coded this up myself, but if you are typing at this level of detail I think you can use `OrdKeySpine` here. It uses less memory / compute when you statically know that the value type is `()`.

Reply:
You might also be able to use `count().map(|(key,aggs)| /* pack */)` if you want; I think this fuses the two which is fine, but I wanted to call out the perhaps simpler design.

Reply:
So if I'm reading the compiler errors and the differential docs correctly this can't be a OrdKeySpine because the second argument defines the output type which is an arrangement of (Row, Row)

I don't fully understand the decomposition between the arrange*, trace* and Ord* stuff in differential so I might be misreading but just dropping a OrdKeySpine here doesnt work

Reply:
That being said, I stole a chunk of this operator from `ReduceCollation` I can do whatever no strong feelings

Reply:
Ah, I misunderstood/miscommunicated. That argument is the *output* arrangement, and does need to be what it is. The *input* arrangement, unspecified but something you could wire up, can be `OrdKeySpine`.

Reply:
Ok I just pushed a commit that makes the edit you wanted both here and in the accumulable case

Reply:
In which case, I think things are probably good to go?